src/packet/bwe/arrival_group.rs

```diff
@@ -0,0 +1,309 @@
+    /// * [`Belongs::NewGroup`] if the packet is the start of a new group and thus wasn't added.
+    ///                         The caller should create a new group and add this packet to the
+    ///                         group.
+    /// * [`Belongs::Skipped`] if the packet was skipped i.e. because it was out of order.
+    fn add_packet(&mut self, packet: AckedPacket) -> Belongs {
+        let belongs = self.belongs_to_group(&packet);
+
+        if belongs == Belongs::NewGroup || belongs == Belongs::Skipped {
+            return belongs;
+        }
+
+        self.remote_recv_time = Some(packet.remote_recv_time);
+
+        if self.first_local_send_time.is_none() {
+            self.first_local_send_time = Some(packet.local_send_time);
+            self.first_seq_no = Some(packet.seq_no);
+        }
+
+        self.local_send_time = self.local_send_time.max(Some(packet.local_send_time));
+        self.size += 1;
+        self.last_seq_no = Some(packet.seq_no);
+
+        Belongs::Belongs
+    }
+
+    fn belongs_to_group(&self, packet: &AckedPacket) -> Belongs {
+        let Some(first_local_send_time) = self.first_local_send_time else {
+            // Start of the group
+            return Belongs::Belongs;
+        };
+
+        let Some(send_diff) = packet
+            .local_send_time
+            .checked_duration_since(first_local_send_time) else {
+                warn!("Out of order send");
+                // Out of order
+                return Belongs::Skipped;
+        };
+
+        if send_diff < BURST_TIME_INTERVAL {
+            // Sent within the same burst interval
+            return Belongs::Belongs;
+        }
+
+        let inter_arrival_time = packet
+            .remote_recv_time
+            .checked_duration_since(self.remote_recv_time());
+
+        let Some(inter_arrival_time) = inter_arrival_time else {
+            warn!("Out of order arrival");
+            return Belongs::Skipped;
+        };
+
+        let inter_group_delay_variation = inter_arrival_time.as_secs_f64()
+            - (packet.local_send_time - self.local_send_time()).as_secs_f64();
+
+        return (inter_group_delay_variation < 0.0 && inter_arrival_time < BURST_TIME_INTERVAL)
+            .into();
+    }
+
+    /// Calculate the inter group delay variation between self and a subsequent group.
+    pub(super) fn inter_group_delay_variation(&self, other: &Self) -> f64 {
+        let arrival_variation = self.arrival_variation(&other);
+        let departure_variation = self.departure_variation(&other);
+
+        assert!(arrival_variation >= 0.0);
+
+        let result = arrival_variation - departure_variation;
+        let first_seq_no = self.first_seq_no.unwrap();
+        let last_seq_no = self.last_seq_no.unwrap();
+        trace!("Delay variation for group({first_seq_no}..={last_seq_no}. {result} = {arrival_variation} - {departure_variation}");
+
+        result
+    }
+
+    pub(super) fn departure_variation(&self, other: &Self) -> f64 {
```

Comment:
In one place you have "send_time" and in the other "departure".  I'd suggest picking one.  And if I had to pick, I'd say make this "local_send_time_variation".

Reply:
Yup, I raised this inconsistency elsewhere.

The spec uses "departure variation" but libWebRTC uses "send diff" or "send delta". Then the question is still "local send" vs just "send", i.e. is it worth using the "local" and "remote" prefixes for clarity